prog/modes/initmodes.F90

```diff
@@ -326,4 +335,55 @@ subroutine readGeometry(geonode, geo)
 
   end subroutine readGeometry
 
+
+  !> Reads atomic masses from input file, overwriting those from the SK files
+  subroutine getInputMasses(node, geo, masses)
+
+    !> relevant node of input data
+    type(fnode), pointer :: node
+
+    !> geometry object, which contains atomic species information
+    type(TGeometry), intent(in) :: geo
+
+    !> masses to be returned
+    real(dp), allocatable, intent(out) :: masses(:)
+
+    type(fnode), pointer :: child, child2, child3, val
+    type(fnodeList), pointer :: children
+    integer, allocatable :: pTmpI1(:)
+    type(string) :: buffer, modifier
+    real(dp) :: rTmp
+    integer :: ii, jj, iAt
+
+    call getChildValue(node, "Masses", val, "", child=child, allowEmptyValue=.true.,&
+        & dummyValue=.true., list=.true.)
+
+    ! Read individual atom specifications
+    call getChildren(child, "Mass", children)
+    if (getLength(children) == 0) then
+      return
+    end if
+
+    allocate(masses(geo%nAtom))
+    masses(:) = -1.0_dp
+    do ii = 1, getLength(children)
+      call getItem1(children, ii, child2)
+      call getChildValue(child2, "Atoms", buffer, child=child3, multiple=.true.)
+      call convAtomRangeToInt(char(buffer), geo%speciesNames, geo%species, child3, pTmpI1)
+      call getChildValue(child2, "MassPerAtom", rTmp, modifier=modifier, child=child)
+      call convertByMul(char(modifier), massUnits, child, rTmp)
+      do jj = 1, size(pTmpI1)
+        iAt = pTmpI1(jj)
+        if (masses(iAt) >= 0.0_dp) then
+          call detailedWarning(child3, "Previous setting for the mass  of atom" // i2c(iAt) //&
+              & " overwritten")
+        end if
+        masses(iAt) = rTmp
+      end do
+      deallocate(pTmpI1)
```

Comment:
A minor node: You don't necessarily need to deallocate here, as it will be automatically deallocated on entry to `convAtomRangeToInt` since the dummy argument is `allocatable, intent(out)`.